website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+---
+authors:
+  - jackkoenig
+tags: [design patterns]
+slug: arbiter-type-class
+description: Type classes and hardware design.
+---
+
+# Type Classes and Hardware Design
+
+When giving talks about Chisel, I always say that Chisel is intended for writing _reusable hardware generators_.
+The thesis is that, by virtue of being embedded in Scala, Chisel can take advantage of powerful programming language features to make hardware design easier.
+How to actually do this; however, is usually left as an exercise for the audience.
+
+In this blog post, we'll explore how one might use the *type class* pattern to build Chisel generators.
+
+## A Specialized Arbiter for a Specific Protocol
+
+Let's first build a simple fixed-priority arbiter that works for a specific kind of client:
+
+```scala
+class ArbiterClient extends Bundle {
+  val request = Bool()
+  val grant = Flipped(Bool())
+}
+
+class PriorityArbiter(nClients: Int) extends Module {
+  val clients = IO(Vec(nClients, Flipped(new ArbiterClient)))
+
+  val requests = clients.map(_.request)
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    client.grant := grant
+  }
+}
+```
+
+<!-- truncate -->
```

Comment:
what does this do?

Reply:
It's a directive to tell Docusaurus to only show what is above in the "collapsed" view on the webiste. I didn't use this for the only other blog post: https://www.chisel-lang.org/blog

If I put that between the first and second paragraphs, then it would only show the first paragraph and display a "Read More" button to expand. For an example, you can see the "Read More" behavior on the Docusaurus blog here: https://docusaurus.io/blog

Reply:
And if you don't have it, Docusaurus prints a warning (and since this post is long, the Blog page looks a bit silly)